File: Minsk/Parser.cs
using System.Collections.Immutable;

namespace Minsk;

public class Parser
{
    private readonly ImmutableArray<SyntaxToken> _tokens;
    private int _position;

    public Parser(string text)
    {
        _tokens = new Lexer(text)
            .Select(tok => tok!)
            .Where(tok => tok.Kind != SyntaxKind.BadToken && tok.Kind != SyntaxKind.WhitespaceToken)
            .ToImmutableArray();
    }

    private SyntaxToken Peek(int offset)
    {
        var index = _position + offset;
        if (index >= _tokens.Length)
        {
            return new SyntaxToken(SyntaxKind.EndOfFileToken, "", _tokens.Last().Position, null);
        }

        return _tokens[index];
    }

    private SyntaxToken Current => Peek(0);

    private SyntaxToken NextToken()
    {
        var current = Current;
        _position++;
        return current;
    }

    private SyntaxToken MatchToken(SyntaxKind kind)
    {
        if (Current is { } token && token.Kind == kind)
        {
            return NextToken();
        }

        return new SyntaxToken(kind, "", Current.Position, null);
    }

    public ExpressionSyntax Parse()
    {
        var left = ParsePrimaryExpression();

        while (Current is { Kind: SyntaxKind.PlusToken or SyntaxKind.MinusToken })
        {
            var operatorToken = NextToken();
            var right = ParsePrimaryExpression();
            left = new BinaryExpressionSyntax(left, operatorToken, right);
        }

        return left;
    }

    private ExpressionSyntax ParsePrimaryExpression()
    {
        var numberToken = MatchToken(SyntaxKind.NumberToken);
        return new LiteralExpressionSyntax(numberToken);
    }
}

File: kt/Minsk/src/main/kotlin/dev/phytolizer/minsk/Parser.kt
package dev.phytolizer.minsk

class Parser(text: String) {
    private val tokens: List<SyntaxToken>
    private var position = 0

    init {
        val tempTokens = mutableListOf<SyntaxToken>()
        for (token in Lexer(text)) {
            if (token.kind != SyntaxKind.BadToken && token.kind != SyntaxKind.WhitespaceToken) {
                tempTokens.add(token)
            }
        }
        tempTokens.add(SyntaxToken(SyntaxKind.EndOfFileToken, tempTokens.lastOrNull()?.position ?: 0, "", null))
        tokens = tempTokens
    }

    private fun peek(offset: Int): SyntaxToken {
        val index = position + offset
        if (index >= tokens.size) {
            return tokens.last()
        }

        return tokens[index]
    }

    private val current
        get() = peek(0)

    private fun nextToken(): SyntaxToken {
        val curr = current
        position += 1
        return curr
    }

    private fun matchToken(kind: SyntaxKind): SyntaxToken {
        if (current.kind == kind) {
            return nextToken()
        }

        return SyntaxToken(kind, current.position, "", null)
    }

    fun parse(): ExpressionSyntax {
        return parseExpression()
    }

    private fun parseExpression(): ExpressionSyntax {
        return parseTerm()
    }

    private fun parseTerm(): ExpressionSyntax {
        var left = parseFactor()

        while (current.kind == SyntaxKind.PlusToken || current.kind == SyntaxKind.MinusToken) {
            val operatorToken = nextToken()
            val right = parseFactor()
            left = BinaryExpressionSyntax(left, operatorToken, right)
        }

        return left
    }

    private fun parseFactor(): ExpressionSyntax {
        var left = parsePrimaryExpression()

        while (current.kind == SyntaxKind.StarToken || current.kind == SyntaxKind.SlashToken) {
            val operatorToken = nextToken()
            val right = parsePrimaryExpression()
            left = BinaryExpressionSyntax(left, operatorToken, right)
        }

        return left
    }

    private fun parsePrimaryExpression(): ExpressionSyntax {
        val numberToken = matchToken(SyntaxKind.NumberToken)
        return LiteralExpressionSyntax(numberToken)
    }
}
